guest/rust/examples/basics/procedural_generation/src/client.rs

```diff
@@ -0,0 +1,367 @@
+    sampler::create(&sampler::Descriptor {
+        address_mode_u: sampler::AddressMode::ClampToEdge,
+        address_mode_v: sampler::AddressMode::ClampToEdge,
+        address_mode_w: sampler::AddressMode::ClampToEdge,
+        mag_filter: sampler::FilterMode::Nearest,
+        min_filter: sampler::FilterMode::Nearest,
+        mipmap_filter: sampler::FilterMode::Nearest,
+    })
+}
+
+fn default_linear_sampler() -> ProceduralSamplerHandle {
+    sampler::create(&sampler::Descriptor {
+        address_mode_u: sampler::AddressMode::ClampToEdge,
+        address_mode_v: sampler::AddressMode::ClampToEdge,
+        address_mode_w: sampler::AddressMode::ClampToEdge,
+        mag_filter: sampler::FilterMode::Linear,
+        min_filter: sampler::FilterMode::Linear,
+        mipmap_filter: sampler::FilterMode::Linear,
+    })
+}
+
+fn make_procedural<BaseColorFn, NormalFn, MetallicRoughnessFn, SamplerFn>(
+    world_translation: Vec3,
+    base_color_fn: BaseColorFn,
+    normal_fn: NormalFn,
+    metallic_roughness_fn: MetallicRoughnessFn,
+    sampler_fn: SamplerFn,
+    transparent: bool,
+) where
+    BaseColorFn: FnMut(f32, f32) -> [u8; 4],
+    NormalFn: FnMut(f32, f32) -> [u8; 4],
+    MetallicRoughnessFn: FnMut(f32, f32) -> [u8; 4],
+    SamplerFn: FnOnce() -> ProceduralSamplerHandle,
+{
+    let mut vertices = vec![];
+    let mut indices = vec![];
+    for y in 0..=RESOLUTION_Y {
+        for x in 0..=RESOLUTION_X {
+            let px = SIZE_X * (x as f32) / (RESOLUTION_X as f32);
+            let py = SIZE_Y * (y as f32) / (RESOLUTION_Y as f32);
+            let pz = WAVE_AMPLITUDE * f32::sin(WAVE_FREQUENCY * px);
+            let u = (x as f32) / (RESOLUTION_X as f32);
+            let v = (y as f32) / (RESOLUTION_Y as f32);
+            vertices.push(mesh::Vertex {
+                position: vec3(px, py, pz) + vec3(-0.5 * SIZE_X, -0.5 * SIZE_Y, 0.0),
+                normal: vec3(0.0, 0.0, 1.0),
+                tangent: vec3(1.0, 0.0, 0.0),
+                texcoord0: vec2(u, v),
+            });
+        }
+    }
+    for y in 0..RESOLUTION_Y {
+        for x in 0..RESOLUTION_X {
+            let i0 = x + y * (RESOLUTION_X + 1);
+            let i1 = (x + 1) + y * (RESOLUTION_X + 1);
+            let i2 = x + (y + 1) * (RESOLUTION_X + 1);
+            let i3 = (x + 1) + (y + 1) * (RESOLUTION_X + 1);
+            indices.extend([i0, i1, i2]);
+            indices.extend([i1, i3, i2]);
+        }
+    }
+    for triangle in indices.chunks_exact(3) {
+        let [i0, i1, i2]: [_; 3] = triangle.try_into().unwrap();
+        let p0 = vertices[i0 as usize].position;
+        let p1 = vertices[i1 as usize].position;
+        let p2 = vertices[i2 as usize].position;
+        let n01 = (p1 - p0).normalize();
+        let n02 = (p2 - p0).normalize();
+        let n = n01.cross(n02).normalize();
+        vertices[i0 as usize].normal += n;
+        vertices[i1 as usize].normal += n;
+        vertices[i2 as usize].normal += n;
```

Comment:
Consider adding an option to `mesh::Descriptor` to have the normals automatically calculated by the host

Reply:
Yeah let's do this later. I think we could also add automatically computed UVs (some trivial projection) and tangents (from UVs) too.